Update R CRAN packages

glennpj opened the pull request:
This PR updates R CRAN packages. It also ensures that spack built
packages that correspond to R recommended packages are specified in the
dependency list. This is needed for #12015.

@adamjstewart 
Please add the WIP label.

Thanks.

glennpj commented:
I see now that using `spack versions` to check for updates does not work as expected. It is only checking the contents of `list_url` so it can not catch the current version from the remote versions.

glennpj commented:
I am also seeing that packages were added without specifying the full dependencies. I am not referring to the recommended packages but dependencies for packages that are not even in spack yet. I will catch those in additional PRs later.

glennpj commented:
Filed an issue for the `spack versions/checksum` problem, #12020.

glennpj commented:
I think some of the missing dependencies are due to new versions needing additional packages. I will be skipping those in this PR and creating new PRs, as mentioned above.

I have also come across a case where there is a dependency on one of the R base packages. Those are always built with R and are often probably the reason a spec might say R≥3.0.0, etc. Those will block activation of packages. So, in addition to adding dependencies for recommended packages I will also remove dependencies for base packages. 😒

glennpj commented:
I am bumping dependency versions where specified. It is possible that some of those may not work with older versions of packages but I think we can deal with those as they come up. This is definitely making me question whether there is a better way to manage these packages in spack.

glennpj commented:
@adamjstewart I am now at the stage of adding new packages to satisfy dependencies. It is amazing how much the R package dependencies can change even across what would appear to be minor versions. In any event, there will be ~40 new R packages needed to satisfy new dependencies. Would you prefer that I submit individual PRs for each of those, or individual commits into this PR?

Also, I had to abandon the idea of trying to avoid duplication with #12005 due to needing to satisfy dependencies. I guess one of these PRs will need to be rebased after the other goes in. @codeandkey has a list in https://github.com/spack/spack/pull/12005#issuecomment-513855452 and there is some overlap. It is not a huge number of packages.

adamjstewart commented:
You can add them all as a single commit (or multiple) to this PR if you want. Let me know when things are ready to review and I'll take a look!

glennpj commented:
@adamjstewart I think this PR is ready for another set of eyes. I verified that all of the new packages build. I also verified that all of the packages successfully concretize.

To sum up what was done.

- make sure that any "recommended" packages were listed as dependencies to satisfy #12015 
- update versions
- add new dependencies where needed
- set additional constraints on dependencies and versions where needed
new dependencies could be introduced even with minor version changes. Likewise, some dependencies dropped off with version changes
- Fixed over constraints, such as `depends_on('r@3.5.0:3.5.9')` when the package actually could work with other versions of R 
- added new packages to satisfy the dependency list of the existing packages with the new versions
- switched the packages to use `cloud.r-project` to provide better responsiveness than downloading from `cran.r-project`

adamjstewart commented:
> switched the packages to use cloud.r-project to provide better responsiveness than downloading from cran.r-project

Can you update https://spack.readthedocs.io/en/latest/build_systems/rpackage.html to use this suggestion?

adamjstewart commented on var/spack/repos/builtin/packages/r-acepack/package.py:
```diff
@@ -9,8 +9,8 @@
     """ACE and AVAS for Selecting Multiple Regression Transformations."""
 
-    homepage = "https://CRAN.R-project.org/package=acepack"
+    homepage = "https://cloud.R-project.org/package=acepack"
```
Switch to lowercase R

glennpj replied:
Okay; my script was just looking for cran/cloud, but I should be able to add this.

glennpj replied:
This should be fixed with the latest commit.

adamjstewart commented on var/spack/repos/builtin/packages/r-acepack/package.py:
```diff
@@ -9,8 +9,8 @@
 
-    homepage = "https://CRAN.R-project.org/package=acepack"
+    homepage = "https://cloud.R-project.org/package=acepack"
     url      = "https://cran.rstudio.com/src/contrib/acepack_1.4.1.tar.gz"
```
Switch to https://cloud.r-project.org/src/contrib/acepack_1.4.1.tar.gz

glennpj replied:
I did not replace any of the rstudio references but I can do another pass.

glennpj replied:
This is fixed with the latest commit.

adamjstewart commented on var/spack/repos/builtin/packages/r-ade4/package.py:
```diff
@@ -11,9 +11,12 @@ class RAde4(RPackage):
+    version('1.7-11', sha256='4ccd799ae99bd625840b866a697c4a48adb751660470bf0d6cf9207b1927a572')
     version('1.7-6', '63401ca369677538c96c3d7b75b3f4a1')
 
     depends_on('r@2.10:')
```
Probably `type=('build', 'run')`

glennpj replied:
Since spack does not have a version of R < 3.1 I think I will remove any specs < 3.1.2. For the ones above that I will add the type.

adamjstewart replied:
I like keeping the constraint just in case someone adds an old version, or more importantly, if someone uses a system installation as an external package and it is old.

glennpj replied:
Got it.

glennpj replied:
This is fixed with the latest commit.

adamjstewart commented on var/spack/repos/builtin/packages/r-aer/package.py:
```diff
@@ -12,16 +12,17 @@ class RAer(RPackage):
-    url      = "https://cran.r-project.org/src/contrib/AER_1.2-5.tar.gz"
+    homepage = "https://cloud.r-project.org/web/packages/AER/index.html"
+    url      = "https://cloud.r-project.org/src/contrib/AER_1.2-5.tar.gz"
     list_url = "https://cran.rstudio.com/src/contrib/Archive/AER"
```
Switch to https://cloud.r-project.org/src/contrib/Archive/AER

glennpj replied:
This is fixed with the latest commit.

adamjstewart commented on var/spack/repos/builtin/packages/r-aer/package.py:
```diff
@@ -12,16 +12,17 @@ class RAer(RPackage):
 
-    homepage = "https://cran.r-project.org/web/packages/AER/index.html"
-    url      = "https://cran.r-project.org/src/contrib/AER_1.2-5.tar.gz"
+    homepage = "https://cloud.r-project.org/web/packages/AER/index.html"
```
Switch to https://cloud.r-project.org/package=AER

glennpj replied:
This is fixed with the latest commit.

glennpj commented:
@adamjstewart said:
>Make sure homepage, url, and list_url all use cloud.r-project.org. 

The home page is frequently not on the r-project site. For those that do not have a separate home page though, then yea, it should be consistent.

glennpj commented:
I am also finding some older packages that did not have a `list_url` setting so I am catching those as well.

adamjstewart commented:
I'll try to take another look tonight.

adamjstewart commented on var/spack/repos/builtin/packages/r-bfast/package.py:
```diff
@@ -9,13 +9,12 @@
 
     version('1.5.7', sha256='01585fe8944d05ebdb13795214077bc1365f0c0372e2a1f7edb914356dace558')
 
-    depends_on('r@2.15.0:', type=('build', 'run'))
```
Are you sure this shouldn't be here? According to https://cloud.r-project.org/web/packages/bfast/index.html it says: R (≥ 2.15.0)

adamjstewart commented on var/spack/repos/builtin/packages/r-adegenet/package.py:
```diff
@@ -18,13 +18,15 @@ class RAdegenet(RPackage):
+    version('2.1.0', sha256='7ee44061002b41164bbc09256307ab02e536f4f2ac03f36c7dc8f85f6af4639a')
     version('2.0.1', 'ecb1220ce7c9affaba2987bc7f38adda')
 
-    depends_on('r@2.14:')
```
According to https://cran.r-project.org/web/packages/adegenet/index.html, it says: R (≥ 2.14)

adamjstewart commented on var/spack/repos/builtin/packages/r-aer/package.py:
```diff
@@ -12,16 +12,17 @@ class RAer(RPackage):
+    version('1.2-6', sha256='653c3a2d253819e0ce8c2cf12cff2ab222bf3d19dbf382b7c4b4c3d762469474')
     version('1.2-5', '419df9dc8ee6e5edd79678fee06719ae')
 
-    depends_on('r@2.13.0:', type=('build', 'run'))
```
According to https://cloud.r-project.org/package=AER, it actually requires R (≥ 3.0.0)

adamjstewart commented on var/spack/repos/builtin/packages/r-caroline/package.py:
```diff
@@ -10,10 +10,8 @@ class RCaroline(RPackage):
 
     version('0.7.6', sha256='e7ba948f7d87f091b498dd0eec2ca4fdad7af4e2bbb67e0945c2f0d3f2eadda9')
-
-    depends_on('r@1.8.0:', type=('build', 'run'))
```
According to https://cloud.r-project.org/package=caroline it requires R (≥ 1.8.0)

glennpj replied:
I think I had taken those out some time back and forgot about them. I will make another pass through these and catch those although I really can not see anyone using versions of R that old.

glennpj replied:
@adamjstewart Also, there are some packages that did not have the R version specified even for newer versions of R. It looks like I missed some of those. Not sure if I will have another commit this evening but let me get those in before you spend too much time on that aspect.

glennpj replied:
@adamjstewart Sorry this took so long but I had trouble finding blocks of time to work on it. The packages have R versions specified now.

adamjstewart commented:
I'll try to take a look at this by tomorrow.

adamjstewart commented on var/spack/repos/builtin/packages/r-adabag/package.py:
```diff
@@ -9,12 +9,15 @@
     version('4.1', '2e019f053d49f62ebb3b1697bbb50afa')
 
     depends_on('r-rpart', type=('build', 'run'))
     depends_on('r-mlbench', type=('build', 'run'))
```
This dependency is suggested, not required. I would remove this.

glennpj replied:
I too noticed that some packages had included "suggested" packages. My first instinct was to remove them but then I was not sure if the package author had simply overdid the dependency specs or if there was a good reason to include those. I left them in to avoid potentially breaking something. My feeling is that the dependencies were simply over stated so the risk to breakage is probably small.

glennpj replied:
This actually depends on mlbench up to v4.1. I Made that change.

adamjstewart commented on var/spack/repos/builtin/packages/r-aer/package.py:
```diff
@@ -12,16 +12,18 @@ class RAer(RPackage):
 
-    depends_on('r@2.13.0:', type=('build', 'run'))
+    depends_on('r@3.0.0:', type=('build', 'run'))
     depends_on('r-car@2.10-19:', type=('build', 'run'))
```
```suggestion
    depends_on('r-car@2.10-19:', type=('build', 'run'))
```
```suggestion
    depends_on('r-car@2.0-19:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-ape/package.py:
```diff
@@ -24,9 +24,10 @@ class RApe(RPackage):
+    version('5.3', sha256='08b0df134c523feb00a86896d1aa2a43f0f0dab20a53bc6b5d6268d867988b23')
     version('5.2', sha256='27eb02856c130d59de6e06276be4981709923756319e465a7f2d4756d4f46415')
     version('5.1', sha256='b7d5dca66881638227a40aa59533904aa5efe0f4b867851b248e8f948a01a26e')
     version('5.0', sha256='c32ed22e350b3d7c7ef3de9334155ab1f3086922b5ec9a1643897cae7abda960')
```
`r-rcpp@0.12.0:`

glennpj replied:
Made this change.

adamjstewart commented on var/spack/repos/builtin/packages/r-bibtex/package.py:
```diff
@@ -11,9 +11,9 @@ class RBibtex(RPackage):
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/bibtex/"
 
     version('0.4.2', sha256='1f06ab3660c940405230ad16ff6e4ba38d4418a59cd9b16d78a4349f8b488372')
 
```
The `r-testthat` dependency is suggested, not required. It should be removed.

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-bindrcpp/package.py:
```diff
@@ -11,12 +11,12 @@ class RBindrcpp(RPackage):
-    depends_on('r-rcpp', type=('build', 'run'))
-    depends_on('r-bindr', type=('build', 'run'))
+    depends_on('r-rcpp@0.12.16:', type=('build', 'run'))
+    depends_on('r-bindr@1.1.1:', type=('build', 'run'))
```
```suggestion
    depends_on('r-bindr@0.1.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-blob/package.py:
```diff
@@ -12,9 +12,14 @@ class RBlob(RPackage):
+    depends_on('r-tibble', when='@:1.1.0', type=('build', 'run'))
+    depends_on('r-prettyunits', when='@1.2.0:', type=('build', 'run'))
+    depends_on('r-rlang', when='@1.2.0:', type=('build', 'run'))
+    depends_on('r-vctrs', when='@1.2.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r-vctrs@0.2.0:', when='@1.2.0:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-bookdown/package.py:
```diff
@@ -10,13 +10,16 @@ class RBookdown(RPackage):
+    depends_on('r-knitr@1.22:', type=('build', 'run'))
     depends_on('r-htmltools@0.3.6:', type=('build', 'run'))
+    depends_on('r-xfun@0.6:', when='@0.12:', type=('build', 'run'))
+    depends_on('r-tinytex@0.12:', when='@0.12:', type=('build', 'run'))
```
One more requirement:

> SystemRequirements:	Pandoc (>= 1.17.2)

glennpj replied:
This is actually kind of a big one. Pandoc is a Haskell package, which has a really odd build system, ultimately requiring a binary version of GHC to build GHC. There are binaries of pandoc available for MacOS and Linux and I wonder if that would be a better way to go. In any event, I will put his on my list of things to work on, but since the pandoc dependency has been missing from the beginning, and it will take a while to sort out how to best get that into spack, I do not think this should hold up this PR.

glennpj replied:
Indeed, the issue with GHC/pandoc was discussed extensively in #1408.

adamjstewart replied:
At the very least, you should add a commented out dependency and a TODO note.

glennpj replied:
That might be best for now. After reading through #1408 I am inclined to go with a binary install of pandoc. Given that this dependency was never in place I wonder if this package actually works, at least in the case where pandoc is not installed as a system package. Is there a way to mark a package broken in spack? I suppose something could be done with a `conflicts` statement.

glennpj replied:
@adamjstewart I am looking into creating a pandoc package with the binaries that are being produced from version 2.0+. I would like for this to work on Linux and Mac but I am not sure how to handle the checksums in `version()`. There could be two downloads, depending on the platform:

1. pandoc-2.7.3-linux.tar.gz
1. pandoc-2.7.3-macOS.zip

I know there would be a `url_for_version` needed but how do I handle the two different checksums in `version()` for the different downloads?

adamjstewart replied:
There's currently no way to do this. The best you can do is something like:
```python
import platform

class Foo(FooPackage):
    if platform.system() == 'darwin':
        # macOS versions
        version('1.2.3', sha256='...')
    elif platform.system() == 'linux':
        # Linux versions
        version('1.2.3', sha256='...')
```

glennpj replied:
@adamjstewart I added a pandoc package.

glennpj replied:
Added dependency on pandoc to r-bookdown.

adamjstewart commented on var/spack/repos/builtin/packages/r-callr/package.py:
```diff
@@ -12,13 +12,14 @@ class RCallr(RPackage):
     version('3.0.0', sha256='e36361086c65660a6ecbbc09b5ecfcddee6b59caf75e983e48b21d3b8defabe7')
     version('1.0.0', 'd9af99bb95696310fa1e5d1cb7166c91')
 
     depends_on('r-base64enc', type=('build', 'run'), when='@3.0.0:')
```
This dependency doesn't seem to be needed.

glennpj replied:
I removed it.

adamjstewart commented on var/spack/repos/builtin/packages/r-car/package.py:
```diff
@@ -11,14 +11,24 @@ class RCar(RPackage):
 
+    depends_on('r@3.2.0:', when='@:3.0-2', type=('build', 'run'))
+    depends_on('r@3.5.0:', when='@3.0-3:', type=('build', 'run'))
+    depends_on('r-cardata', when='@3.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r-cardata', when='@3.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r-cardata@3.0-0:', when='@3.0:', type=('build', 'run'))
```

glennpj commented:
> > switched the packages to use cloud.r-project to provide better responsiveness than downloading from cran.r-project
> 
> Can you update https://spack.readthedocs.io/en/latest/build_systems/rpackage.html to use this suggestion?

@adamjstewart I was going to work on that today but I see that you beat me to it in #12471. Sorry I did not get to that sooner.

glennpj commented:
@adamjstewart I will try to take some time each day and make commits for your recommendations. Would you prefer that I hold off on pushing those until you are done with your review?

glennpj commented:
@adamjstewart A question came up on the Spack Google Groups site regarding making java a variant. I think that is probably best done as a separate PR as opposed to further modifications here and in #12015 but I wanted to get your take.

adamjstewart commented:
> Would you prefer that I hold off on pushing those until you are done with your review?

You can push any time.

adamjstewart commented:
> @adamjstewart A question came up on the Spack Google Groups site regarding making java a variant. I think that is probably best done as a separate PR as opposed to further modifications here and in #12015 but I wanted to get your take.

I agree, let's do a separate PR for R +/- Java stuff.

adamjstewart commented on var/spack/repos/builtin/packages/r-crayon/package.py:
```diff
@@ -13,9 +13,9 @@ class RCrayon(RPackage):
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/crayon"
 
     version('1.3.4', '77c7c2906c59a3141306d86c89ffc7d3')
     version('1.3.2', 'fe29c6204d2d6ff4c2f9d107a03d0cb9')
```
Depends:	R (≥ 3.0.2)

glennpj replied:
@adamjstewart I am not seeing the R constraint listed on the CRAN page or the DESCRIPTION file for r-crayon.

adamjstewart replied:
Oh, maybe that's because it has the wrong homepage!

glennpj replied:
That would do it. For some reason that got changed in 050b753.

glennpj replied:
Fixed.

adamjstewart commented on var/spack/repos/builtin/packages/r-crul/package.py:
```diff
@@ -9,10 +9,11 @@
+    version('0.8.4', sha256='dbd950ad3b68402e5a5955615b1abcb5c9bdc846c93aa25f96a7a58913d04c8b')
     version('0.7.4', sha256='c963dd666ae3fc89b661ce19fce2fa19a16fc3825e1502105cae98ceb92c6014')
 
     depends_on('r-curl@3.3:', type=('build', 'run'))
```
Missing dependency on `r-jsonlite`

glennpj replied:
Ah, jsonlite was in the "suggests" list until version 0.8.4, which was added to CRAN on 08/02/2019. I had checked dependencies against version 0.7.4 but then I did another update round after that, and it moved from "suggests" to "imports" with that latest version. I added the dependency for r-jsonlite, with the version constraint.

adamjstewart commented on var/spack/repos/builtin/packages/r-dendextend/package.py:
```diff
@@ -9,15 +9,17 @@
+    depends_on('r@3.0.0:', type=('build', 'run'))
     depends_on('r-magrittr@1.0.1:', type=('build', 'run'))
     depends_on('r-ggplot2', type=('build', 'run'))
     depends_on('r-fpc', type=('build', 'run'))
```
This is either no longer a dependency or never was.

glennpj replied:
It was a dependency up until version 1.10.0. I added that constraint.

adamjstewart commented on var/spack/repos/builtin/packages/r-digest/package.py:
```diff
@@ -25,11 +25,13 @@ class RDigest(RPackage):
     version('0.6.9',  '48048ce6c466bdb124716e45ba4a0e83')
 
-    depends_on('r@2.4.1:')
+    depends_on('r@2.4.1:', type=('build', 'run'))
```
```suggestion
    depends_on('r@3.1.0:', type=('build', 'run'))
```

glennpj replied:
Another where the version constraint changed that I missed. Switched to R-3.1.0 with version 0.6.16.

adamjstewart commented on var/spack/repos/builtin/packages/r-doparallel/package.py:
```diff
@@ -10,12 +10,14 @@ class RDoparallel(RPackage):
     version('1.0.10', 'd9fbde8f315d98d055483ee3493c9b43')
 
+    depends_on('r@2.14.0:', type=('build', 'run'))
     depends_on('r-foreach', type=('build', 'run'))
```
```suggestion
    depends_on('r-foreach@1.2.0:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-doparallel/package.py:
```diff
@@ -10,12 +10,14 @@ class RDoparallel(RPackage):
 
+    depends_on('r@2.14.0:', type=('build', 'run'))
     depends_on('r-foreach', type=('build', 'run'))
     depends_on('r-iterators', type=('build', 'run'))
```
```suggestion
    depends_on('r-iterators@1.0.0:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-dt/package.py:
```diff
@@ -25,3 +28,4 @@ class RDt(RPackage):
     depends_on('r-htmlwidgets@1.3:', type=('build', 'run'))
```
Missing dependency:

> jsonlite (≥ 0.9.16)

glennpj replied:
Another case where `jsonlite` moved from "suggests" to "imports" in an update that I did after checking dependencies. Added the constraint when version is 0.8 and above.

adamjstewart commented on var/spack/repos/builtin/packages/r-dygraphs/package.py:
```diff
@@ -13,13 +13,16 @@ class RDygraphs(RPackage):
-    depends_on('r-htmlwidgets', type=('build', 'run'))
-    depends_on('r-zoo', type=('build', 'run'))
-    depends_on('r-xts', type=('build', 'run'))
+    depends_on('r-htmlwidgets@0.6:', type=('build', 'run'))
```
Missing dependency:

>  htmltools (≥ 0.3.5)

glennpj replied:
Fixed.

adamjstewart commented on var/spack/repos/builtin/packages/r-ergm/package.py:
```diff
@@ -12,16 +12,22 @@ class RErgm(RPackage):
-    depends_on('r-matrix', type=('build', 'run'))
-    depends_on('r-lpsolve', type=('build', 'run'))
-    depends_on('r-mass', type=('build', 'run'))
+    depends_on('r-robustbase@0.93-5:', type=('build', 'run'))
```
```suggestion
    depends_on('r-robustbase@0.93.5:', type=('build', 'run'))
```

glennpj replied:
Nothing to do here.

adamjstewart commented on var/spack/repos/builtin/packages/r-ergm/package.py:
```diff
@@ -12,16 +12,22 @@ class RErgm(RPackage):
-    depends_on('r-lpsolve', type=('build', 'run'))
-    depends_on('r-mass', type=('build', 'run'))
+    depends_on('r-robustbase@0.93-5:', type=('build', 'run'))
+    depends_on('r-coda@0.19-2:', type=('build', 'run'))
```
```suggestion
    depends_on('r-coda@0.19.2:', type=('build', 'run'))
```

glennpj replied:
@adamjstewart Both coda and robustbase, as well as many other R packages, use a '-' in the version string.

adamjstewart replied:
As far as Spack is concerned, it doesn't make a difference. As long as the CRAN page for coda or the CRAN page for ergm uses a dash, that's fine. The trust dependency below is the only one that is definitely incorrect.

glennpj replied:
Yep; even packages that use a dash are sometimes referred to with a '.', which is confusing. I think they actually redirect to the canonical one in the end.

glennpj replied:
Nothing to do here.

adamjstewart commented on var/spack/repos/builtin/packages/r-ergm/package.py:
```diff
@@ -12,16 +12,22 @@ class RErgm(RPackage):
-    depends_on('r-mass', type=('build', 'run'))
+    depends_on('r-robustbase@0.93-5:', type=('build', 'run'))
+    depends_on('r-coda@0.19-2:', type=('build', 'run'))
+    depends_on('r-trust@01-7:', type=('build', 'run'))
```
```suggestion
    depends_on('r-trust@0.1.7:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-ergm/package.py:
```diff
@@ -12,16 +12,22 @@ class RErgm(RPackage):
+    depends_on('r-robustbase@0.93-5:', type=('build', 'run'))
+    depends_on('r-coda@0.19-2:', type=('build', 'run'))
+    depends_on('r-trust@01-7:', type=('build', 'run'))
+    depends_on('r-matrix@1.2-17:', type=('build', 'run'))
```
```suggestion
    depends_on('r-matrix@1.2.17:', type=('build', 'run'))
```

glennpj replied:
Nothing to do here.

adamjstewart commented on var/spack/repos/builtin/packages/r-ergm/package.py:
```diff
@@ -12,16 +12,22 @@ class RErgm(RPackage):
+    depends_on('r-trust@01-7:', type=('build', 'run'))
+    depends_on('r-matrix@1.2-17:', type=('build', 'run'))
+    depends_on('r-lpsolve@5.6.13:', type=('build', 'run'))
+    depends_on('r-mass@7.3-51.4:', type=('build', 'run'))
```
```suggestion
    depends_on('r-mass@7.3.51.4:', type=('build', 'run'))
```

glennpj replied:
Nothing to do here.

adamjstewart commented on var/spack/repos/builtin/packages/r-ergm/package.py:
```diff
@@ -12,16 +12,22 @@ class RErgm(RPackage):
+    depends_on('r-matrix@1.2-17:', type=('build', 'run'))
+    depends_on('r-lpsolve@5.6.13:', type=('build', 'run'))
+    depends_on('r-mass@7.3-51.4:', type=('build', 'run'))
     depends_on('r-statnet-common@3.3:', type=('build', 'run'))
```
```suggestion
    depends_on('r-statnet-common@4.3.0:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-ergm/package.py:
```diff
@@ -12,16 +12,22 @@ class RErgm(RPackage):
+    depends_on('r-purrr@0.3.2:', when='@3.10.0:', type=('build', 'run'))
+    depends_on('r-rlang@0.3.4:', when='@3.10.0:', type=('build', 'run'))
+    depends_on('r-tibble@2.1.1:', when='@3.10.0:', type=('build', 'run'))
+    depends_on('r-dplyr@0.8.0.1:', when='@3.10.0:', type=('build', 'run'))
```
Missing a system dependency:

> SystemRequirements:	OpenMPI

You can probably use any generic MPI

glennpj replied:
@adamjstewart This is an interesting one and I would appreciate your input. The OpenMPI system requirement was added for version 3.10.4, the latest version. However, looking at the documentation, and the source, it is not really a dependency but more of, if it detects that it is running under MPI, ie., started with mpirun, then it will use it. However, in order to use MPI it would also need to have the Rmpi (r-rmpi) and the doSNOW (r-dosnow) packages, and by extension, the snow (r-snow) packages installed. None of those actually list MPI as a dependency for installation of the respective R packages. Furthermore, those are not listed as dependencies for r-ergm, although Rmpi is in the suggested list.

If one follows the MPI example in the docs it is clear what packages one would need to install and load to use MPI, and nothing in r-ergm has to be built beforehand for that support, so it does seem to be unnecessary from a packaging perspective. I suppose that a variant could be used to ensure that the MPI related R packages can be optionally installed with r-ergm, but the two variants would be the exact same package, just one installing extra packages along the way. I lean towards leaving out the MPI dependency but I am not sure if there have been other cases like this in Spack that set precedent.

glennpj replied:
Based on a related comment about the r-snow package, it seems like leaving the OpenMPI (MPI) dependency out is the best course to take.

glennpj replied:
I added a comment about using MPI with this package.

adamjstewart replied:
I agree with this decision.

adamjstewart commented on var/spack/repos/builtin/packages/r-expm/package.py:
```diff
@@ -10,8 +10,12 @@ class RExpm(RPackage):
-    homepage = "http://R-Forge.R-project.org/projects/expm"
-    url      = "https://cran.r-project.org/src/contrib/expm_0.999-2.tar.gz"
-    list_url = "https://cran.r-project.org/src/contrib/Archive/expm"
+    homepage = "http://R-Forge.r-project.org/projects/expm"
```
```suggestion
    homepage = "http://r-forge.r-project.org/projects/expm"
```

adamjstewart commented on var/spack/repos/builtin/packages/r-factoextra/package.py:
```diff
@@ -11,18 +11,19 @@ class RFactoextra(RPackage):
     version('1.0.4', 'aa4c81ca610f17fdee0c9f3379e35429')
 
-    depends_on('r@3.1.0:')
+    depends_on('r@3.1.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r@3.1.2:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-ff/package.py:
```diff
@@ -11,9 +11,11 @@ class RFf(RPackage):
     version('2.2-13', '4adf8840d24cb6e2927a3ef885c86fcd')
 
+    depends_on('r@2.10.1:', type=('build', 'run'))
     depends_on('r-bit', type=('build', 'run'))
```
```suggestion
    depends_on('r-bit@1.1-13:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-fnn/package.py:
```diff
@@ -11,10 +11,12 @@ class RFnn(RPackage):
+    version('1.1.2.2', sha256='b51a60fbbeff58c48cc90c2023c48972d5082d68efd02284c17ccd9820986326')
     version('1.1',   '8ba8f5b8be271785593e13eae7b8c393')
     version('1.0',   'e9a47dc69d1ba55165be0877b8443fe0')
     version('0.6-4', '1c105df9763ceb7b13989cdbcb542fcc')
```
The last 2 dependencies are suggestions, not required. Remove them

glennpj replied:
They have been removed.

adamjstewart commented on var/spack/repos/builtin/packages/r-formatr/package.py:
```diff
@@ -14,13 +14,16 @@ class RFormatr(RPackage):
     version('1.4', '98b9b64b2785b35f9df403e1aab6c73c')
 
+    depends_on('r@3.0.2:', type=('build', 'run'))
     depends_on('r-codetools', type=('build', 'run'))
```
All of these dependencies are suggestions, not required. Remove them

glennpj replied:
Those have been removed.

adamjstewart commented on var/spack/repos/builtin/packages/r-fpc/package.py:
```diff
@@ -10,20 +10,22 @@ class RFpc(RPackage):
-    # depends_on('r-class', type=('build', 'run'))
+    depends_on('r-class', type=('build', 'run'))
     depends_on('r-diptest', type=('build', 'run'))
     depends_on('r-mvtnorm', type=('build', 'run'))
```
Suggested, not required. Remove

glennpj replied:
It was an "imports"  up until version 2.2-2. Added that constraint.

adamjstewart commented on var/spack/repos/builtin/packages/r-fs/package.py:
```diff
@@ -0,0 +1,20 @@
+    version('1.3.1', sha256='d6934dca8f835d8173e3fb9fd4d5e2740c8c04348dd2bcc57df1b711facb46bc')
+
+    depends_on('r@3.1:', type=('build', 'run'))
+    depends_on('r-rcpp', type=('build', 'run'))
```
Missing dependency:
```python
depends_on('gmake', type='build')
```

glennpj replied:
Added gmake dependency.

adamjstewart commented on var/spack/repos/builtin/packages/r-future-apply/package.py:
```diff
@@ -0,0 +1,25 @@
+    version('1.3.0', sha256='6374eca49bb81e05c013509c8e324cf9c5d023f9f8217b29ce7b7e12025ca371')
+
+    depends_on('r@3.2.0:', type=('build', 'run'))
+    depends_on('r-future@1.13.0:', type=('build', 'run'))
```
Missing dependency:

> Imports:	globals (≥ 0.12.4)

glennpj replied:
No idea how I missed this one.

glennpj commented on var/spack/repos/builtin/packages/r-clipr/package.py:
```diff
@@ -11,9 +11,11 @@ class RClipr(RPackage):
+    version('0.5.0', sha256='fd303f8b7f29badcdf490bb2d579acdfc4f4e1aa9c90ac77ab9d05ce3d053dbf')
     version('0.4.0', '4012a31eb3b7a36bd3bac00f916e56a7')
 
     depends_on('r-rstudioapi', type=('build', 'run'))
```
Done.

glennpj commented on var/spack/repos/builtin/packages/r-construct/package.py:
```diff
@@ -10,9 +10,9 @@ class RConstruct(RPackage):
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/conStruct"
 
     version('1.0.3', sha256='b449c133a944ad05a28f84f312ed4ccbc1574c4659aa09c678618d2ae9008310')
 
```
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-getoptlong/package.py:
```diff
@@ -12,12 +12,14 @@ class RGetoptlong(RPackage):
     depends_on('r-rjson', type=('build', 'run'))
-    depends_on('r-globaloptions', type=('build', 'run'))
+    depends_on('r-globaloptions@0.1.0:', type=('build', 'run'))
     depends_on('perl')
```
Missing dependency on `perl-getopt-long`: https://metacpan.org/pod/Getopt::Long

glennpj replied:
The getopt::long module is installed as part of Perl so this is already satisfied with the perl dependency.

glennpj replied:
I will add a comment in the package file about the Getopt::Long dependency.

adamjstewart commented on var/spack/repos/builtin/packages/r-ggally/package.py:
```diff
@@ -14,15 +14,18 @@ class RGgally(RPackage):
     depends_on('r-rcolorbrewer', type=('build', 'run'))
-    depends_on('r-reshape', type=('build', 'run'))
+    depends_on('r-reshape@0.8.5:', type=('build', 'run'))
+    depends_on('r-rlang', when='@1.4.0:', type=('build', 'run'))
```
> SystemRequirements:	openssl

glennpj replied:
Needed for version 1.4.0 and above; added.

adamjstewart commented on var/spack/repos/builtin/packages/r-ggrepel/package.py:
```diff
@@ -10,12 +10,14 @@ class RGgrepel(RPackage):
 
-    depends_on('r@3.0.0:')
+    depends_on('r@3.0.0:', type=('build', 'run'))
     depends_on('r-ggplot2@2.0.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r-ggplot2@2.2.0:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-globaloptions/package.py:
```diff
@@ -10,12 +10,15 @@ class RGlobaloptions(RPackage):
+    depends_on('r@3.3.0:', when='@0.0.13:', type=('build', 'run'))
     depends_on('r-testthat', type=('build', 'run'))
     depends_on('r-knitr', type=('build', 'run'))
     depends_on('r-markdown', type=('build', 'run'))
```
These 3 are suggested, not required. Removed

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-haven/package.py:
```diff
@@ -11,13 +11,17 @@ class RHaven(RPackage):
+    depends_on('r-readr@0.1.0:', type=('build', 'run'))
     depends_on('r-hms', type=('build', 'run'))
     depends_on('r-tibble', type=('build', 'run'))
     depends_on('r-forcats', type=('build', 'run'))
```
```suggestion
    depends_on('r-forcats@0.2.0:', type=('build', 'run'))
```

glennpj replied:
This was done.

adamjstewart commented on var/spack/repos/builtin/packages/r-haven/package.py:
```diff
@@ -11,13 +11,17 @@ class RHaven(RPackage):
+    version('2.1.1', sha256='90bcb4e7f24960e7aa3e15c06b95cd897f08de149cec43fd8ba110b14526068a')
+    version('2.1.0', sha256='c0a1cf1b039549fb3ad833f9644ed3f142790236ad755d2ee7bd3d8109e3ae74')
     version('1.1.0', '8edd4b7683f8c36b5bb68582ac1b8733')
 
```
> SystemRequirements:	GNU make

glennpj replied:
Added gmake.

glennpj commented on var/spack/repos/builtin/packages/r-gdalutils/package.py:
```diff
@@ -21,3 +21,4 @@ class RGdalutils(RPackage):
     depends_on('r-utils', type=('build', 'run'))
```
This is something that I have on my TODO list that I was going to bring up after the dust settled in this PR. Now is a good time though. There is no `r-r-utils` package in Spack. The `r-utils` package is actually for `R.utils`. This is confusing because there is a base package named `utils` and I saw some cases where that was used when the spec was actually for the base package. The bottom line is that the `r-utils` package should be called `r-r-utils` to be consistent with other package naming. The same is true for the `r-oo` package, it should be called `r-r-oo`, as it corresponds to R.oo. At least that one does not have the same name as a base package.

adamjstewart replied:
Sounds fine by me, I think it's few enough packages that I don't really care. I'll let you decide.

glennpj replied:
Is there a process for renaming Spack packages, other than the git operations?

glennpj commented on var/spack/repos/builtin/packages/r-gdata/package.py:
```diff
@@ -25,11 +25,12 @@ class RGdata(RPackage):
     version('2.17.0', 'c716b663b9dc16ad8cafe6acc781a75f')
 
+    depends_on('r@2.3.0:', type=('build', 'run'))
     depends_on('r-gtools', type=('build', 'run'))
```
Added perl dependency.

adamjstewart commented on var/spack/repos/builtin/packages/r-htmlwidgets/package.py:
```diff
@@ -11,14 +11,15 @@ class RHtmlwidgets(RPackage):
-    homepage = "https://cran.r-project.org/package=htmlTable"
-    url      = "https://cran.rstudio.com/src/contrib/htmlwidgets_0.9.tar.gz"
-    list_url = "https://cran.r-project.org/src/contrib/Archive/htmlwidgets"
+    homepage = "https://cloud.r-project.org/package=htmlTable"
```
Wrong homepage

glennpj replied:
Fixed.

adamjstewart commented on var/spack/repos/builtin/packages/r-httpuv/package.py:
```diff
@@ -16,12 +16,15 @@ class RHttpuv(RPackage):
     depends_on('r-rcpp@0.11.0:', type=('build', 'run'))
+    depends_on('r-r6', when='@1.5.0:', type=('build', 'run'))
+    depends_on('r-promises', when='@1.5.0:', type=('build', 'run'))
+    depends_on('r-later@0.8.0:', when='@1.5.0:', type=('build', 'run'))
```
> SystemRequirements:	GNU make

glennpj replied:
Added.

adamjstewart commented on var/spack/repos/builtin/packages/r-log4r/package.py:
```diff
@@ -10,10 +10,11 @@ class RLog4r(RPackage):
+    version('0.3.0', sha256='8e5d0221298410e48bee9d9a983a23e1834ce88592f9d931471bfdb05f37a691')
     version('0.2', 'f3fcb7b1f48526c6543b2e00e278ff65')
 
     depends_on('r-testthat', type=('build', 'run'))
```
Remove this dependency

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-loo/package.py:
```diff
@@ -11,8 +11,8 @@ class RLoo(RPackage):
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/loo"
 
     version('2.1.0', sha256='1bf4a1ef85d151577ff96d4cf2a29c9ef24370b0b1eb08c70dcf45884350e87d')
 
```
> SystemRequirements:	pandoc (>= 1.12.3), pandoc-citeproc

glennpj replied:
Looking at the PR for the original package, #11448, it was determined that pandoc was not really a dependency for r-loo.

glennpj replied:
I went ahead and added the dependency for pandoc.

adamjstewart commented on var/spack/repos/builtin/packages/r-lubridate/package.py:
```diff
@@ -14,12 +14,15 @@ class RLubridate(RPackage):
-    depends_on('r-rcpp@0.11:', type=('build', 'run'))
+    depends_on('r@3.0.0:', type=('build', 'run'))
+    depends_on('r-rcpp@0.12.13:', type=('build', 'run'))
     depends_on('r-stringr', type=('build', 'run'))
```
> SystemRequirements:	A system with zoneinfo data (e.g. /usr/share/zoneinfo) as well as a recent-enough C++11 compiler (such as g++-4.8 or later). On Windows the zoneinfo included with R is used.

I wonder if zoneinfo is something we can package

glennpj replied:
It would seem that most systems would have zoneinfo data.

glennpj replied:
The zoneinfo data can be gotten from [IANA — Time Zone Database](https://www.iana.org/time-zones), so it could be packaged. I think this is something that is better handled at the system level though, unless there are a significant numbers of systems that use Spack that would not have the zoneinfo data installed.

adamjstewart commented on var/spack/repos/builtin/packages/r-maldiquant/package.py:
```diff
@@ -17,11 +17,14 @@ class RMaldiquant(RPackage):
 
+    depends_on('r@3.2.0:', type=('build', 'run'))
     depends_on('r-knitr', type=('build', 'run'))
     depends_on('r-testthat', type=('build', 'run'))
```
These 2 are suggested dependencies, remove them.

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-methodss3/package.py:
```diff
@@ -17,8 +17,10 @@ class RMethodss3(RPackage):
     a cross-platform package implemented in pure R that generates
```
Change package name to `r-r-methodss3`

glennpj replied:
Yep, already have this on my TODO list

adamjstewart commented on var/spack/repos/builtin/packages/r-minqa/package.py:
```diff
@@ -11,9 +11,9 @@ class RMinqa(RPackage):
     version('1.2.4', 'bcaae4fdba60a33528f2116e2fd51105')
 
-    depends_on('r-rcpp', type=('build', 'run'))
+    depends_on('r-rcpp@0.9.10:', type=('build', 'run'))
```
> SystemRequirements:	GNU make

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-mlbench/package.py:
```diff
@@ -10,12 +10,11 @@ class RMlbench(RPackage):
-    depends_on('r@2.10:')
-
+    depends_on('r@2.10:', type=('build', 'run'))
     depends_on('r-lattice', type=('build', 'run'))
```
Suggested, remove

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-ncdf4/package.py:
```diff
@@ -23,9 +23,11 @@ class RNcdf4(RPackage):
+    version('1.16', sha256='edd5731a805bbece3a8f6132c87c356deafc272351e1dd07256ca00574949253')
     version('1.15', 'cd60dadbae3be31371e1ed40ddeb420a')
 
     depends_on('netcdf')
```
```suggestion
    depends_on('netcdf@4.1:')
```

adamjstewart commented on var/spack/repos/builtin/packages/r-nloptr/package.py:
```diff
@@ -16,10 +16,11 @@ class RNloptr(RPackage):
+    version('1.2.1', sha256='1f86e33ecde6c3b0d2098c47591a9cd0fa41fb973ebf5145859677492730df97')
     version('1.0.4', 'f2775dfb4f7f5552d46937a04c062b0d')
 
     depends_on('r-testthat', type=('build', 'run'))
```
Not required, remove

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-oo/package.py:
```diff
@@ -16,9 +16,11 @@ class ROo(RPackage):
     standard S3 classes without any tricks."""
```
`r-oo` -> `r-r-oo`

adamjstewart commented on var/spack/repos/builtin/packages/r-oo/package.py:
```diff
@@ -16,9 +16,11 @@ class ROo(RPackage):
 
-    depends_on('r-methodss3', type=('build', 'run'))
+    depends_on('r@2.13.0:', type=('build', 'run'))
+    depends_on('r-methodss3@1.7.1:', type=('build', 'run'))
```
```suggestion
    depends_on('r-r-methodss3@1.7.1:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-packrat/package.py:
```diff
@@ -11,8 +11,12 @@ class RPackrat(RPackage):
     version('0.4.8-1', '14e82feba55fcda923396282fc490038')
     version('0.4.7-1', '80c2413269b292ade163a70ba5053e84')
+
+    depends_on('r@3.3.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r@3.0.0:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-paramhelpers/package.py:
```diff
@@ -14,10 +14,14 @@ class RParamhelpers(RPackage):
     version('1.10', '36e9060488ebd484d62cd991a4693332')
 
     depends_on('r-bbmisc@1.10:', type=('build', 'run'))
     depends_on('r-checkmate@1.8.1:', type=('build', 'run'))
```
```suggestion
    depends_on('r-checkmate@1.8.2:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-pbdzmq/package.py:
```diff
@@ -17,10 +17,14 @@ class RPbdzmq(RPackage):
+    depends_on('r@3.0.0:', when='@:0.2-5', type=('build', 'run'))
+    depends_on('r@3.2.0:', when='@0.2-6:', type=('build', 'run'))
     depends_on('r-r6', type=('build', 'run'))
     depends_on('zeromq')
```
```suggestion
    depends_on('zeromq@4.0.4:')
```

adamjstewart commented on var/spack/repos/builtin/packages/r-pbdzmq/package.py:
```diff
@@ -17,10 +17,14 @@ class RPbdzmq(RPackage):
 
+    depends_on('r@3.0.0:', when='@:0.2-5', type=('build', 'run'))
+    depends_on('r@3.2.0:', when='@0.2-6:', type=('build', 'run'))
     depends_on('r-r6', type=('build', 'run'))
```
I don't see this dependency listed anywhere

glennpj replied:
It was needed up until version 0.2-6. I added that constraint.

adamjstewart commented on var/spack/repos/builtin/packages/r-pkgbuild/package.py:
```diff
@@ -9,18 +9,19 @@
 
     depends_on('r@3.1:', type=('build', 'run'))
-    depends_on('r-callr', type=('build', 'run'))
+    depends_on('r-callr@2.0.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r-callr@3.2.0:', type=('build', 'run'))
```

glennpj replied:
This changed to r-callr@3.2.0: with version 1.0.4, which was an update after I did the check. Added additional constraints.

adamjstewart commented on var/spack/repos/builtin/packages/r-pkgmaker/package.py:
```diff
@@ -16,8 +16,8 @@ class RPkgmaker(RPackage):
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/pkgmaker"
 
     version('0.27', sha256='17a289d8f596ba5637b07077b3bff22411a2c2263c0b7de59fe848666555ec6a')
     version('0.22', '73a0c6d3e84c6dadf3de7582ef7e88a4')
```
Has a bunch of suggested dependencies to remove

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-plogr/package.py:
```diff
@@ -11,9 +11,11 @@ class RPlogr(RPackage):
     version('0.2.0', sha256='0e63ba2e1f624005fe25c67cdd403636a912e063d682eca07f2f1d65e9870d29')
     version('0.1-1', '5ee46ed21b5c571d02900219098517c0')
+
+    depends_on('r-rcpp', when='@0.2.0:', type=('build', 'run'))
```
Suggested, not required, remove

glennpj replied:
Removed

adamjstewart commented on var/spack/repos/builtin/packages/r-rda/package.py:
```diff
@@ -10,8 +10,12 @@ class RRda(RPackage):
-    homepage = "https://cran.r-project.org/web/packages/rda/index.html"
-    url      = "https://cran.r-project.org/src/contrib/rda_1.0.2-1.tar.gz"
-    list_url = "https://cran.r-project.org/src/contrib/Archive/rda"
+    homepage = "https://cloud.r-project.org/web/packages/rda/index.html"
```
```suggestion
    homepage = "https://cloud.r-project.org/package=rda"
```

adamjstewart commented on var/spack/repos/builtin/packages/r-readr/package.py:
```diff
@@ -12,14 +12,19 @@ class RReadr(RPackage):
     depends_on('r-r6', type=('build', 'run'))
     depends_on('r-bh', type=('build', 'run'))
+    depends_on('r-clipr', when='@1.2.0:', type=('build', 'run'))
+    depends_on('r-crayon', when='@1.3.1:', type=('build', 'run'))
```
> SystemRequirements:	GNU make

glennpj replied:
Added.

adamjstewart commented on var/spack/repos/builtin/packages/r-rematch/package.py:
```diff
@@ -11,8 +11,9 @@ class RRematch(RPackage):
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/rematch"
 
     version('1.0.1', '5271666295e232931f21499522489dd3')
 
```
These dependencies aren't required

glennpj replied:
Removed.

adamjstewart commented on var/spack/repos/builtin/packages/r-remotes/package.py:
```diff
@@ -0,0 +1,21 @@
+
+    version('2.1.0', sha256='8944c8f6fc9f0cd0ca04d6cf1221b716eee08facef9f4b4c4d91d0346d6d68a7')
+
+    depends_on('r@2.6.0:', type=('build', 'run'))
```
```suggestion
    depends_on('r@3.0.0:', type=('build', 'run'))
```

glennpj commented on var/spack/repos/builtin/packages/r-hoardr/package.py:
```diff
@@ -9,9 +9,9 @@
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/hoardr/"
 
     version('0.5.2', sha256='819113f0e25da105f120a676b5173872a4144f2f6f354cad14b35f898e76dc54')
 
```
Removed the suggested dependencies.

glennpj commented on var/spack/repos/builtin/packages/r-raster/package.py:
```diff
@@ -11,11 +11,13 @@ class RRaster(RPackage):
+    version('2.9-23', sha256='90aaec9e3b1e3e6015d9993ea7491e008f2f71990f8abb8610f979c4e28b38af')
+    version('2.9-22', sha256='8107d95f1aa85cea801c8101c6aa391becfef4b5b915d9bc7a323531fee26128')
     version('2.5-8', '2a7db931c74d50516e82d04687c0a577')
 
```
Added.

glennpj commented on var/spack/repos/builtin/packages/r-rcppprogress/package.py:
```diff
@@ -11,14 +11,15 @@ class RRcppprogress(RPackage):
     version('0.1',   '34afefe0580ca42b6353533fe758d5bf')
 
-    depends_on('r@3.4.0:3.4.9')
     depends_on('r-rcpp', type=('build', 'run'))
```
It was up until version 0.4. Added the constraint.

glennpj commented on var/spack/repos/builtin/packages/r-rcurl/package.py:
```diff
@@ -18,11 +18,13 @@ class RRcurl(RPackage):
 
+    depends_on('r@3.0.0:', type=('build', 'run'))
     depends_on('r-bitops', type=('build', 'run'))
     depends_on('curl')
```
Added.

glennpj commented on var/spack/repos/builtin/packages/r-rcppcnpy/package.py:
```diff
@@ -10,9 +10,10 @@ class RRcppcnpy(RPackage):
+    version('0.2.10', sha256='77d6fbc86520a08da40d44c0b82767099f8f719ca95870d91efff1a9cab1ab9c')
     version('0.2.9', '7f63354d15928b6716830c2975b3baf0')
 
     depends_on('r@3.1.0:', type=('build', 'run'))
```
There is a copy of the cnpy source included in the distribution that has been modified to some degree for Rcpp, and that is what is built. Therefore, the cnpy dependency is not necessary. I will remove the cnpy dependency.

adamjstewart commented on var/spack/repos/builtin/packages/r-rrcov/package.py:
```diff
@@ -9,9 +9,9 @@
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/rrcov"
 
     version('1.4-7', sha256='cbd08ccce8b583a2f88946a3267c8fc494ee2b44ba749b9296a6e3d818f6f293')
 
```
Last 2 are suggested, remove

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-rsnns/package.py:
```diff
@@ -17,9 +17,12 @@ class RRsnns(RPackage):
     version('0.4-7', 'ade7736611c456effb5f72e0ce0a1e6f')
 
+    depends_on('r@2.10.0:', type=('build', 'run'))
     depends_on('r-rcpp', type=('build', 'run'))
```
```suggestion
    depends_on('r-rcpp@0.8.5:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-rsqlite/package.py:
```diff
@@ -11,17 +11,20 @@ class RRsqlite(RPackage):
     depends_on('r-memoise', type=('build', 'run'))
     depends_on('r-pkgconfig', type=('build', 'run'))
-    depends_on('r-rcpp', type=('build', 'run'))
+    depends_on('r-rcpp@1.12.7:', type=('build', 'run'))
```
```suggestion
    depends_on('r-rcpp@0.12.7:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-rstan/package.py:
```diff
@@ -18,14 +18,17 @@ class RRstan(RPackage):
+    depends_on('r-ggplot2@2.0.0:', type=('build', 'run'))
     depends_on('r-stanheaders', type=('build', 'run'))
     depends_on('r-inline', type=('build', 'run'))
     depends_on('r-gridextra', type=('build', 'run'))
```
```suggestion
    depends_on('r-gridextra@2.0.0:', type=('build', 'run'))
```

adamjstewart replied:
Also, Rcpp (≥ 0.12.0), BH (≥ 1.69.0)

glennpj replied:
Those, and some other changes were made. There were duplicate dependency specifications and unneeded when conditions for some.

adamjstewart commented on var/spack/repos/builtin/packages/r-rstan/package.py:
```diff
@@ -37,5 +40,5 @@ class RRstan(RPackage):
     depends_on('r-pkgbuild', type=('build', 'run'), when='@2.18:')
-    depends_on('r-stanheaders@2.18.0:', type=('build', 'run'), when='@2.18:')
+    depends_on('r-stanheaders@2.18.1:', type=('build', 'run'), when='@2.18:')
     conflicts('%gcc@:4.9', when='@2.18:')
```
> SystemRequirements:	GNU make, pandoc

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-rstantools/package.py:
```diff
@@ -10,7 +10,7 @@ class RRstantools(RPackage):
+    url      = "https://cloud.r-project.org/src/contrib/rstantools_1.5.1.tar.gz"
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/rstantools"
 
     version('1.5.1', sha256='5cab16c132c12e84bd08e18cd6ef25ba39d67a04ce61015fc4490659c7cfb485')
```
> SystemRequirements:	pandoc

glennpj replied:
Added.

adamjstewart commented on var/spack/repos/builtin/packages/r-runit/package.py:
```diff
@@ -9,9 +9,9 @@
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/RUnit/"
 
     version('0.4.32', sha256='23a393059989000734898685d0d5509ece219879713eb09083f7707f167f81f1')
 
```
XML is suggested, remove

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-stanheaders/package.py:
```diff
@@ -24,9 +24,12 @@ class RStanheaders(RPackage):
     version('2.17.1', '11d8770277dd18e563852852633c6c25')
     version('2.10.0-2', '9d09b1e9278f08768f7a988ad9082d57')
+
+    depends_on('r@3.4.0:', when='@2.18.0:', type=('build', 'run'))
```
> SystemRequirements:	pandoc

glennpj replied:
Added.

adamjstewart commented on var/spack/repos/builtin/packages/r-stringi/package.py:
```diff
@@ -18,12 +18,15 @@ class RStringi(RPackage):
     version('1.1.1', '32b919ee3fa8474530c4942962a6d8d9')
 
+    depends_on('r@2.14:', type=('build', 'run'))
     depends_on('icu4c')
```
```suggestion
    depends_on('icu4c@52:')
```

adamjstewart commented on var/spack/repos/builtin/packages/r-stringr/package.py:
```diff
@@ -13,13 +13,17 @@ class RStringr(RPackage):
     version('1.0.0', '5ca977c90351f78b1b888b379114a7b4')
 
+    depends_on('r@3.1:', type=('build', 'run'))
     depends_on('r-stringi', type=('build', 'run'))
```
```suggestion
    depends_on('r-stringi@1.1.7:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-survey/package.py:
```diff
@@ -16,7 +16,19 @@ class RSurvey(RPackage):
+    depends_on('r-lattice', when='@3.31:', type=('build', 'run'))
+    depends_on('r-minqa', when='@3.34:', type=('build', 'run'))
+    depends_on('r-numderiv', when='@3.34:', type=('build', 'run'))
+    depends_on('r-mitools', when='@3.36:', type=('build', 'run'))
```
```suggestion
    depends_on('r-mitools@2.4:', when='@3.36:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-utils/package.py:
```diff
@@ -11,9 +11,11 @@ class RUtils(RPackage):
 
-    depends_on('r-oo', type=('build', 'run'))
+    depends_on('r@2.14.0:', type=('build', 'run'))
+    depends_on('r-oo@1.22.0:', type=('build', 'run'))
```
Missing dependency:

> R.methodsS3 (≥ 1.7.1)

adamjstewart commented on var/spack/repos/builtin/packages/r-vfs/package.py:
```diff
@@ -9,9 +9,9 @@
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/VFS"
 
     version('1.0.2', sha256='8ff7e7e13919ff21f10c7c693ef596a2c7b57c7ca37d79278e443ed122a21aad')
 
```
Last 3 are suggested, remove

glennpj replied:
Removed.

adamjstewart commented on var/spack/repos/builtin/packages/r-vgam/package.py:
```diff
@@ -10,16 +10,20 @@ class RVgam(RPackage):
+    depends_on('r@3.1.0:', when='@1.0-2:1.0-3', type=('build', 'run'))
+    depends_on('r@3.4.0:', when='@1.0-4:', type=('build', 'run'))
     depends_on('r-mass', type=('build', 'run'))
     depends_on('r-mgcv', type=('build', 'run'))
```
Last 2 are suggested, remove

glennpj replied:
Removed.

adamjstewart commented on var/spack/repos/builtin/packages/r-xlconnect/package.py:
```diff
@@ -11,11 +11,15 @@ class RXlconnect(RPackage):
 
-    depends_on('r-xlconnectjars', type=('build', 'run'))
+    depends_on('r@2.10.0:', type=('build', 'run'))
+    depends_on('r-xlconnectjars@0.2-15:', type=('build', 'run'))
```
```suggestion
    depends_on('r-xlconnectjars@0.2-15', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/r-xlconnectjars/package.py:
```diff
@@ -10,10 +10,14 @@ class RXlconnectjars(RPackage):
     version('0.2-12', '6984e5140cd1c887c017ef6f88cbba81')
     version('0.2-9', 'e6d6b1acfede26acaa616ee421bd30fb')
 
+    depends_on('r@2.10.0:', type=('build', 'run'))
```
Not listed anywhere

glennpj replied:
Removed.

adamjstewart commented on var/spack/repos/builtin/packages/r-xlsxjars/package.py:
```diff
@@ -10,9 +10,11 @@ class RXlsxjars(RPackage):
     version('0.6.1', '5a1721d5733cb42f3a29e3f353e39166')
 
     depends_on('r-rjava', type=('build', 'run'))
+    depends_on('java@1.6:')
```
Not listed anywhere

glennpj replied:
Removed.

adamjstewart commented on var/spack/repos/builtin/packages/r-xnomial/package.py:
```diff
@@ -11,9 +11,9 @@ class RXnomial(RPackage):
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/XNomial/"
 
     version('1.0.4', sha256='e6237f79d96f02bb30af1cf055ae9f70541abba34ce045a9d4359b5304189dd7')
 
```
knitr is a suggestion, remove

glennpj replied:
Removed.

adamjstewart commented on var/spack/repos/builtin/packages/r-yaml/package.py:
```diff
@@ -10,9 +10,11 @@ class RYaml(RPackage):
-    homepage = "https://cran.r-project.org/web/packages/yaml/index.html"
-    url      = "https://cran.r-project.org/src/contrib/yaml_2.1.13.tar.gz"
-    list_url = "https://cran.r-project.org/src/contrib/Archive/yaml"
+    homepage = "https://cloud.r-project.org/web/packages/yaml/index.html"
```
```suggestion
    homepage = "https://cloud.r-project.org/package=yaml"
```

adamjstewart commented on var/spack/repos/builtin/packages/r-zoo/package.py:
```diff
@@ -14,10 +14,14 @@ class RZoo(RPackage):
 
+    depends_on('r@2.10.0:', when='@:1.8-1', type=('build', 'run'))
+    depends_on('r@3.1.0:', when='@1.8-2:', type=('build', 'run'))
     depends_on('r-lattice', type=('build', 'run'))
```
```suggestion
    depends_on('r-lattice@0.20-27:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/swftools/package.py:
```diff
@@ -0,0 +1,33 @@
+    depends_on('poppler')
+    depends_on('freetype')
+    depends_on('jpeg')
+    depends_on('fftw~mpi')
```
It doesn't work with `fftw+mpi`, or you just don't want to build MPI by default? If it's the latter, you should switch to `depends_on('fftw')` and let people build with whatever they want.

glennpj replied:
@adamjstewart SWFTools does not use MPI. I would imagine that it would work with an FFTW built with MPI but that would pull in unnecessary dependencies. I do see your point though as one could always just add `^fftw~mpi` to the spec when building.

glennpj replied:
I removed the `~mpi` from the fftw dependency spec.

glennpj commented on var/spack/repos/builtin/packages/r-sf/package.py:
```diff
@@ -13,18 +13,21 @@ class RSf(RPackage):
+    depends_on('gdal@2.0.1:')
+    depends_on('geos@3.4.0:')
     # Since PROJ.4 and PROJ.6 have incompatible APIs, I'm assuming
     # that since r-sf supports PROJ.4, it doesn't support PROJ.6
```
It looks to me like support for PROJ6 was added in version 0.7-4. As far as I can tell, PROJ5 works with the older versions. I will add an additional constraint.

glennpj commented on var/spack/repos/builtin/packages/r-reprex/package.py:
```diff
@@ -16,13 +16,20 @@ class RReprex(RPackage):
     depends_on('r-whisker', type=('build', 'run'))
+    depends_on('r-rlang', when='@0.2.0:', type=('build', 'run'))
+    depends_on('r-withr', when='@0.2.0:', type=('build', 'run'))
+    depends_on('r-fs', when='@0.2.1:', type=('build', 'run'))
```
Added.

glennpj commented on var/spack/repos/builtin/packages/r-rgdal/package.py:
```diff
@@ -17,13 +17,15 @@ class RRgdal(RPackage):
+    depends_on('r@3.3.0:', type=('build', 'run'))
+    depends_on('r-sp@1.1-0:', type=('build', 'run'))
     depends_on('gdal')
     depends_on('proj@:5')
```
PROJ6 support was added in version 1.4-2. I adjusted the constraints to reflect that.

glennpj commented on var/spack/repos/builtin/packages/r-rgl/package.py:
```diff
@@ -31,7 +33,7 @@ class RRgl(RPackage):
+    depends_on('r-jsonlite@0.9.20:', type=('build', 'run'))
     depends_on('r-shiny', type=('build', 'run'))
     depends_on('r-magrittr', type=('build', 'run'))
     depends_on('r-crosstalk', type=('build', 'run'), when='@0.99.16:')
```
Added.

glennpj commented on var/spack/repos/builtin/packages/r-rgl/package.py:
```diff
@@ -15,13 +15,15 @@ class RRgl(RPackage):
-    depends_on('r@3.2:3.9')
+    depends_on('r@3.2:', type=('build', 'run'))
     depends_on('zlib', type=('link'))
     depends_on('libpng', type=('link'))
```
Changed.

glennpj commented on var/spack/repos/builtin/packages/r-rgl/package.py:
```diff
@@ -31,7 +33,7 @@ class RRgl(RPackage):
     depends_on('r-htmlwidgets', type=('build', 'run'))
```
Added.

glennpj commented on var/spack/repos/builtin/packages/r-rgl/package.py:
```diff
@@ -15,13 +15,15 @@ class RRgl(RPackage):
+    depends_on('r@3.2:', type=('build', 'run'))
     depends_on('zlib', type=('link'))
     depends_on('libpng', type=('link'))
     depends_on('libx11')
```
I left it in. My guess is that without it will pick up system libraries and headers.

glennpj commented on var/spack/repos/builtin/packages/r-rgooglemaps/package.py:
```diff
@@ -12,11 +12,14 @@ class RRgooglemaps(RPackage):
+    depends_on('r@2.10:', type=('build', 'run'))
     depends_on('r-png', type=('build', 'run'))
-    depends_on('r-rjsonio', type=('build', 'run'))
+    depends_on('r-rjsonio@:1.3.9', type=('build', 'run'))
```
It was required up until version 1.4. There is no 1.3.9 version so changed that to 1.2.0.7.

glennpj commented on var/spack/repos/builtin/packages/r-rhmmer/package.py:
```diff
@@ -9,9 +9,9 @@
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/rhmmer/"
 
     version('0.1.0', sha256='5022cefc9ba335160c1ad8d1b614610ae0ad48287c9fdbaf8f8966149358e520')
 
```
Done.

glennpj commented on var/spack/repos/builtin/packages/r-rjava/package.py:
```diff
@@ -11,10 +11,11 @@ class RRjava(RPackage):
     version('0.9-8', '51ae0d690ceed056ebe7c4be71fc6c7a')
 
+    depends_on('r@2.5:', type=('build', 'run'))
     depends_on('java')
```
Done.

glennpj commented on var/spack/repos/builtin/packages/r-rmarkdown/package.py:
```diff
@@ -11,20 +11,24 @@ class RRmarkdown(RPackage):
+    depends_on('r-stringr@1.2.0:', when='@1.6:', type=('build', 'run'))
+    depends_on('r-catools', when='@:1.7', type=('build', 'run'))
+    depends_on('r-tinytex@0.11:', when='@1.10:', type=('build', 'run'))
+    depends_on('r-xfun', when='@1.13:', type=('build', 'run'))
```
Done.

glennpj commented on var/spack/repos/builtin/packages/r-selectr/package.py:
```diff
@@ -15,12 +15,16 @@ class RSelectr(RPackage):
     version('0.3-1', '7190fcdea1823ad7ef429cab6938e960')
 
+    depends_on('r@3.0:', type=('build', 'run'))
     depends_on('r-testthat', type=('build', 'run'))
```
Done.

adamjstewart commented on var/spack/repos/builtin/packages/r-rgooglemaps/package.py:
```diff
@@ -12,11 +12,14 @@ class RRgooglemaps(RPackage):
+    depends_on('r@2.10:', type=('build', 'run'))
     depends_on('r-png', type=('build', 'run'))
-    depends_on('r-rjsonio', type=('build', 'run'))
+    depends_on('r-rjsonio@:1.2.0.7', type=('build', 'run'))
```
I don't see this dependency listed

glennpj replied:
I resolved this earlier. It is needed for version 1.2.0.7, which is the earliest version in spack. I looked a little further back in versions and found that it first shows up in version 1.2.0.5, but that predates the earliest spack package. I can specify it as 

`depends_on('r-rjsonio@1.2.0.5:1.2.0.7', type=('build', 'run'))`

but it needs to be in there for version 1.2.0.7.

glennpj replied:
Oh wait, I have the spec wrong. It needs to be a `when` condition.

glennpj replied:
@adamjstewart I have it correct now.

`depends_on('r-rjsonio', when='@1.2.0.5:1.2.0.7', type=('build', 'run'))`

glennpj commented on var/spack/repos/builtin/packages/r-seurat/package.py:
```diff
@@ -11,44 +11,75 @@ class RSeurat(RPackage):
+    depends_on('r-rlang', when='@3.0.0:', type=('build', 'run'))
+    depends_on('r-rsvd', when='@3.0.0:', type=('build', 'run'))
+    depends_on('r-scales', when='@3.0.0:', type=('build', 'run'))
+    depends_on('r-sctransform@0.2.0:', when='@3.0.0:', type=('build', 'run'))
```
Those are new dependencies with the new version. None of those are in spack. I updated this package and added the dependencies, as well as 7 new packages to support those new dependencies.

glennpj commented on var/spack/repos/builtin/packages/r-seurat/package.py:
```diff
@@ -11,44 +11,75 @@ class RSeurat(RPackage):
+    depends_on('r-rsvd', when='@3.0.0:', type=('build', 'run'))
+    depends_on('r-scales', when='@3.0.0:', type=('build', 'run'))
+    depends_on('r-sctransform@0.2.0:', when='@3.0.0:', type=('build', 'run'))
+    depends_on('java', when='@:2.3.0', )
```
Fixed.

glennpj commented on var/spack/repos/builtin/packages/r-snow/package.py:
```diff
@@ -9,11 +9,12 @@
 
-    depends_on('r-rmpi', type='run')
     depends_on('r@2.13.1:', type=('build', 'run'))
+    depends_on('r-rmpi', type='run')
```
This is related to my comments regarding the r-ergm package. I think everything is consistent now.

glennpj commented on var/spack/repos/builtin/packages/r-sourcetools/package.py:
```diff
@@ -9,11 +9,13 @@
     version('0.1.5', 'b4d7902ffafd9802e8fbff5ce824bb28')
 
+    depends_on('r@3.0.2:', type=('build', 'run'))
     depends_on('r-testthat', type=('build', 'run'))
```
Done.

glennpj commented on var/spack/repos/builtin/packages/r-taxizedb/package.py:
```diff
@@ -9,9 +9,9 @@
+    list_url = "https://cloud.r-project.org/src/contrib/Archive/taxizedb/"
 
     version('0.1.4', sha256='5a40569a2b5abe56201f112a10220150353412df39b7e8d21ea8698f424cf295')
 
```
Removed.

glennpj commented on var/spack/repos/builtin/packages/r-tclust/package.py:
```diff
@@ -9,17 +9,18 @@
 
-    depends_on('r@3.4.0:3.4.9')
+    depends_on('r@2.12.0:', type=('build', 'run'))
     depends_on('r-cluster', type=('build', 'run'))
```
Done.

glennpj commented on var/spack/repos/builtin/packages/r-threejs/package.py:
```diff
@@ -11,11 +11,16 @@ class RThreejs(RPackage):
+    depends_on('r-htmlwidgets@0.3.2:', type=('build', 'run'))
     depends_on('r-base64enc', type=('build', 'run'))
     depends_on('r-matrix', type=('build', 'run'))
     depends_on('r-jsonlite', type=('build', 'run'))
```
They are needed for version 0.2.2. I added that constraint.

glennpj commented on var/spack/repos/builtin/packages/r-units/package.py:
```diff
@@ -15,9 +15,13 @@ class RUnits(RPackage):
-    depends_on('r-udunits2', type=('build', 'run'))
+    depends_on('r@3.0.2:', type=('build', 'run'))
+    depends_on('r-udunits2@0.13:', when='@:0.5-1', type=('build', 'run'))
+    depends_on('r-rcpp@0.12.10:', type=('build', 'run'))
```
Fixed.

glennpj commented:
The next step is to rename the three packages that need to be renamed.

1. r-utils --> r-r-utils
1. r-oo --> r-r-oo
1. r-methodss3 --> r-r-methodss3

I have built 170 of the r- packages but I would like to build more to see if anything shakes out.

glennpj commented:
I decided to build all of the R CRAN packages, to be complete. The only thing that turned up was that the constraint on mariadb for r-rmysql was too stringent. I committed a change for that. I use an external mariadb but I also tried the build with spack provided mariadb. When doing that I discovered that mariadb-5.5, at least the mariadb.5.5.56 in spack does not build with gcc-9.1.0.  I need to flesh that out a bit more but I think that should be a separate PR specific to mariadb. With that, I _think_ this PR is complete and ready to merge.

adamjstewart commented:
Thanks for all the hard work @glennpj! I think we just need to merge #12005 before we can merge #12015.
